samples/test_azure_event_hubs_receive.py

```diff
@@ -496,6 +496,77 @@ def message_received_callback(message):
+    def data_generator():
+        for i in range(2):
+            msg = uamqp.message.Message(body='Data')
+            # header is only used on received msg, not set on messages being sent
+            msg.application_properties = {'msg_type': 'rich_batch'}
+            msg.properties = uamqp.message.MessageProperties(message_id='richid')
+            msg.footer = {'footerkey':'footervalue'}
+            msg.delivery_annotations = {'deliveryannkey':'deliveryannvalue'}
+            msg.annotations = {'annkey':'annvalue'}
+            yield msg
+
+    uri = "sb://{}/{}".format(live_eventhub_config['hostname'], live_eventhub_config['event_hub'])
+    sas_auth = authentication.SASTokenAuth.from_shared_access_key(
+        uri, live_eventhub_config['key_name'], live_eventhub_config['access_key'])
+
+    target = "amqps://{}/{}/Partitions/0".format(live_eventhub_config['hostname'], live_eventhub_config['event_hub'])
+    send_client = uamqp.SendClient(target, auth=sas_auth, debug=False)
+
+    message = uamqp.message.Message(body='Data')
+    # header is only used on received msg, not set on messages being sent
+    message.application_properties = {'msg_type': 'rich_single'}
+    message.properties = uamqp.message.MessageProperties(message_id='richid')
+    message.footer = {'footerkey':'footervalue'}
+    message.delivery_annotations = {'deliveryannkey':'deliveryannvalue'}
+    message.annotations = {'annkey':'annvalue'}
+    send_client.queue_message(message)
+
+    message_batch = uamqp.message.BatchMessage(data_generator())
+    send_client.queue_message(message_batch)
+    results = send_client.send_all_messages(close_on_done=False)
+    assert not [m for m in results if m == uamqp.constants.MessageState.SendFailed]
+
+    rich_single_received = rich_batch_received = False
+
+    uri = "sb://{}/{}".format(live_eventhub_config['hostname'], live_eventhub_config['event_hub'])
+    sas_auth = authentication.SASTokenAuth.from_shared_access_key(
+        uri, live_eventhub_config['key_name'], live_eventhub_config['access_key'])
+
+    source = "amqps://{}/{}/ConsumerGroups/{}/Partitions/{}".format(
+        live_eventhub_config['hostname'],
+        live_eventhub_config['event_hub'],
+        live_eventhub_config['consumer_group'],
+        0)
+
+    receive_client = uamqp.ReceiveClient(source, auth=sas_auth, timeout=5000, debug=False, prefetch=10)
+    gen = receive_client.receive_messages_iter()
+    for message in gen:
+        if message.application_properties and message.application_properties.get(b'msg_type'):
+            if not rich_single_received:
+                rich_single_received = message.application_properties.get(b'msg_type') == b'rich_single'
+            if not rich_batch_received:
+                rich_batch_received = message.application_properties.get(b'msg_type') == b'rich_batch'
+            assert message.properties.message_id == b'richid'
+            assert message.delivery_annotations
```

Comment:
Is there anything in the spec on whether delivery_annotations _supposed_ to be encoded on outgoing payload? I'm curious whether these are only intended to be written to by the broker? Probably not, but still curious :)

Reply:
there doesn't seem to be any limitation according to the spec, and based on the test, the service supports setting these two properties :)

---

**delivery annotation spec**: http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-delivery-annotations

> The delivery-annotations section is used for delivery-specific non-standard properties at the head of the message. **Delivery annotations convey information from the sending peer to the receiving peer**. If the recipient does not understand the annotation it cannot be acted upon and its effects (such as any implied propagation) cannot be acted upon. Annotations might be specific to one implementation, or common to multiple implementations. The capabilities negotiated on link attach and on the source and target SHOULD be used to establish which annotations a peer supports. A registry of defined annotations and their meanings is maintained [AMQPDELANN]. The symbolic key "rejected" is reserved for the use of communicating error information regarding rejected messages. Any values associated with the "rejected" key MUST be of type error.
If the delivery-annotations section is omitted, it is equivalent to a delivery-annotations section containing an empty map of annotations.

**footer spec**: http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-footer
> The footer section is used for details about the message or delivery which can only be calculated or evaluated once the whole bare message has been constructed or seen (for example message hashes, HMACs, signatures and encryption details).
A registry of defined footers and their meanings is maintained [AMQPFOOTER].